core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java

```diff
@@ -125,13 +132,28 @@ private long skipFromFileChannel(long n) throws IOException {
 
   @Override
   public synchronized void close() throws IOException {
-    fileChannel.close();
-    StorageUtils.dispose(byteBuffer);
+    try {
+      this.cleanable.clean();
+    } catch (UncheckedIOException re) {
+      if (re.getCause() != null) {
+        throw re.getCause();
+      } else {
+        throw re;
+      }
+    }
   }
 
-  @SuppressWarnings("deprecation")
-  @Override
-  protected void finalize() throws IOException {
-    close();
+  private record ResourceCleaner(
+      FileChannel fileChannel,
+      ByteBuffer byteBuffer) implements Runnable {
+    @Override
+    public void run() {
+      try {
+        fileChannel.close();
+        StorageUtils.dispose(byteBuffer);
```

Comment:
This was not done earlier - but do we want to ensure `dispose` is called even if `close` threw an exception ?

Reply:
This is a good question, I think even if `fileChannel.close()` throws an exception, we should try to dispose of the byteBuffer, let's put it in the finally block.